### PR TITLE
Feature/toggle rules in src

### PR DESCRIPTION
**Pull request by donabrams**
So there are a ton more tests to add, and I am definitely having trouble finding the nearest block parent for a comment line. However, I wanted to give you an idea where I was headed.

### CODE REVIEW

**Comment by donabrams** [comment]
This is for https://github.com/sasstools/sass-lint/issues/70

**Comment by Snugug** [comment]
Please include the certificate of origin sign off and let us know when it's ready for comment and review.

**Comment by donabrams** [comment]
DCO 1.1 Signed-off-by: Donald Abrams donabrams@gmail.com

File: tests/ruleToggler.js [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  var file = {
+    'text': fs.readFileSync(filePath),
+    'format': path.extname(filePath).replace('.', ''),
+    'filename': path.basename(filePath)
+  };
+  var ast = groot(file.text, file.format, file.filename);
+  return getToggledRules(ast);
+};
+
+describe.only('rule toggling', function () {
```
Comment (benthemonkey) [inline tests/ruleToggler.js]: Be sure to remove this `.only` so that Coveralls will report an accurate coverage.

Comment (donabrams) [inline tests/ruleToggler.js]: Good catch. I usually add an eslint warning for this.

**Comment by donabrams** [comment]
Ready for comment and review. I'll need to make a separate issue for the doc-disabled-rules rule described in https://github.com/donabrams/sass-lint/blob/ab7e4d8ea2b8f65b09ac97a69e635f7802e758e5/docs/rules/doc-disabled-rules.md

**Comment by benthemonkey** [comment]
Coveralls is unhappy that you didn't test the other branches inside of `sortRange`. I see that it isn't tested because Gonzales-PE doesn't guarantee ordering. Is there any way you could rig the AST to be in the wrong order so we know the code handles that eventuality correctly?

Do you know anything about the performance impact of these extra traversals and checks? Visually there doesn't seem to be any significant impact. Just curious.

**Comment by DanPurdy** [comment]
I'll get this into our sub branch and then work on it a little too, it's looking good though

**Comment by donabrams** [comment]
Performance is split into two independent parts per file, Detection of disable blocks and result filtering.

Detection of disable blocks is done in a single traversal. Time spent will grow linearly with # of comment blocks. No worries there.

The filtering is completed after results are calculated. Filtering time grows at NUM_RESULTS * NUM_GLOBAL_DISABLES + (per Rule: NUM_RULE_RESULTS * NUM_RULE_DISABLES). This is theoretically worrisome, but not in reality: 1) In practice, global disable is usually only done at the top of the page, not many time. 2) If a rule is being ignore multiple times, it's likely it should be ignored once at the top of the page.

**Comment by donabrams** [comment]
Good call @benthemonkey

**Comment by benthemonkey** [comment]
Thanks for the thorough explanation, @donabrams. Glad my suggestion caught some typos.

**Comment by DanPurdy** [comment]
Ok, i'll have a look over this tomorrow, then we should get it into the feature branch we made and throw some more tests at it and maybe some more documentation. Once we're good with that lets get a plan together to release this.

Massive thankyou @donabrams it's looking great and is going to make sass-lint even better! Fingers crossed we may even get this out as part of 1.4.0

**Comment by benthemonkey** [comment]
@donabrams This is being a bit picky, but do you mind going through and updating these commits to be authored by your GitHub account? I think it is helpful to link commits to the GitHub account of the person committing them. Plus, this way you can also show off your rank on the "Contributors" page :smile: 

There's a nice tool to help you do this: https://github.com/davidfokkema/git-rewrite-author

**Comment by donabrams** [comment]
@benthemonkey So git-rewrite-author definitely changed way more than I hoped for. As @DanPurdy pointed out, just adding a secondary email to my account was much easier.

**Comment by benthemonkey** [comment]
Sounds good. I apologize that it caused trouble. When I used that tool everything just worked. This PR looks good to me.

File: lib/ruleToggler.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+      .filter(function (s) {
+        return s.trim().length > 0;
+      });
+    if (!tokens.length) {
+      return;
+    }
+    var first = tokens[0],
+        rules = tokens.slice(1);
+    switch (first) {
+    case 'sass-lint:disable':
```
Comment (DanPurdy) [inline lib/ruleToggler.js]: This indentation isn't flagging up? Looks odd to me but if it's ok with eslint

File: lib/ruleToggler.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  // Sort these toggle stacks so reading them is easier (algorithmically).
+  // Usually already sorted but since it's not guaranteed by the contract with gonzales-pe, ensuring it is.
+  toggledRules.globalEnable.sort(sortRange);
+  Object.keys(toggledRules.ruleEnable).map(function (ruleId) {
+    toggledRules.ruleEnable[ruleId].sort(sortRange);
+  });
+  return toggledRules;
+};
+
+var isBeforeOrSame = function (x, y, x2, y2) {
```
Comment (DanPurdy) [inline lib/ruleToggler.js]: We should move this into the helper functions file and add some jsDoc.

**Comment by DanPurdy** [comment]
I'll carry on with this tomorrow but so far it's looking excellent. For the NPE you're seeing when unable to reliably find a parent block I reckon we should just output an error on that line coming clean that we can't do what they want and suggest using line disables or something. Would that be a fair workaround for now? I'm assuming this is all going to change a little bit within the new gonzales.

**Comment by DanPurdy** [comment]
Hi @donabrams sorry I never got round to this again, been a crazy couple of weeks! I'll get round to this soon and workto get the sass format tests in.

sorry again.

File: lib/ruleToggler.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      }, true);
+    if (!isGloballyEnabled) {
+      return false;
+    }
+    if (!toggledRules.ruleEnable[ruleId]) {
+      return true;
+    }
+    var isRuleEnabled = toggledRules.ruleEnable[ruleId]
+      .reduce(function (acc, toggleRange) {
+        return isBeforeOrSame(line, column, toggleRange[1], toggleRange[2])
```
Comment (donabrams) [inline lib/ruleToggler.js]: Should do ref counting here (ie [disable] [disable] [enable] means it's still disabled.

**Comment by donabrams** [comment]
No worries, I have things I should update on this too... Been swamped past few weeks but I'll get on this soon. Need to:

* Fix parent block does not find to return an error
* Move sort into helpers
* Add ref counting for nested disables when determining enabled so we don't accidentally activate a rule (w/ test cases)

**Comment by gitnik** [comment]
Hey guys, any update on this? :)

**Comment by DanPurdy** [comment]
I think once we've got the gonzales update out of the way this will be the next major priority.

**Comment by gitnik** [comment]
Alright thanks, appreciate the update

**Comment by elengart** [comment]
+1 for this feature... thank you!

**Comment by rcaracaus** [comment]
+1 for this feature

**Comment by dgresko** [comment]
+1! Thanks!

**Comment by jroman00** [comment]
+1

**Comment by DanPurdy** [comment]
Please don't +1 in pull requests it doesn't make anything happen quicker.

**Comment by backflip** [comment]
And it annoys the hell out of everyone having subscribed to the issue. :)

**Comment by necramirez** [comment]
This should be visible in all PRs/issues: https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments

**Comment by ACOSW** [comment]
+1

**Comment by DanPurdy** [comment]
@ACOSW please please please don't use +1's read the thread before commenting...

**Comment by danielbayerlein** [comment]
Are there any chance to merge this important issue?

**Comment by DanPurdy** [comment]
@danielbayerlein unfortunately this is very out of date right now, its one of our next big hurdles to get sorted but merging this right now would probably break a lot of things. We are looking to get it moving though as we know how necessary it is. 👍

**Comment by avinashci** [comment]
I tried the merge and got just one merge conflict on my local on the index.js file. I haven't tested it though.

**Comment by DanPurdy** [comment]
We made massive changes in 1.6 and the AST we are using had massive changes too so we need to go through and review this and throw a lot more tests at it, it will be happening soon though don't worry!

**Comment by DanPurdy** [comment]
Merged this into our test branch for now. Intend to work on this soon. Will open a new PR into develop to see the progress here.

**Comment by DanPurdy** [comment]
Ongoing work in #677 Any work should be a PR into feature/disable-linters branch NOT develop.

Will look to update it all soon.

**Pull request merged**